### PR TITLE
contrib/database/sql: properly annotate DB operations with execution trace tasks

### PR DESCRIPTION
**Pull request by nsrip-dd**
### What does this PR do?

This PR creates execution trace tasks (if the execution tracer is enabled)
before the operations, and provides a way of communicating to the APM tracer
that it doesn't need to create a task for the span. The APM tracer will still
annotate the task with the span ID even if a task was already created.

### Motivation

After #336, spans for database operations are created *after* the operation.
This is necessary because some operations may return ErrSkip if they aren't
supported, and creating spans with those errors is very noisy. But because of
this, execution trace tasks associated with those spans would appear to be only
a few microseconds long since they are created alongside the span and thus only
cover the very end of the operation.

### Describe how to test/QA your changes

This PR adds a unit test which checks that spans get associated tasks with
a realistic duration.

I've tested this locally using a simple program which connects to Postgres and
pg_sleeps:

```
package main

import (
	"context"
	"net/http"
	"os"
	"runtime/trace"

	"github.com/lib/pq"
	sqltrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
)

func main() {
	tracer.Start()
	defer tracer.Stop()

	f, err := os.Create("go.trace")
	if err != nil {
		panic(err)
	}
	defer f.Close()
	if err := trace.Start(f); err != nil {
		panic(err)
	}
	defer trace.Stop()

	sqltrace.Register("pq", &pq.Driver{})
	uri := "host=localhost port=5432 user=postgres password=postgres dbname=postgres sslmode=disable"
	db, err := sqltrace.Open("pq", uri)
	defer db.Close()

	span, ctx := tracer.StartSpanFromContext(context.Background(), "parent")
	_, err = db.ExecContext(ctx, "select pg_sleep(0.5);")
	if err != nil {
		panic(err)
	}
	span.Finish()
}
```

Before:

Screenshot 2023-06-21 at 2 32 21 PM

After:

Screenshot 2023-06-21 at 2 33 02 PM

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality.
- [ ] If this interacts with the agent in a new way, a system test has been added.

### CODE REVIEW

File: contrib/database/sql/exec_trace_test.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/httpmem"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	gotraceui "honnef.co/go/gotraceui/trace"
+)
+
+func TestExecutionTraceAnnotations(t *testing.T) {
+	if trace.IsEnabled() {
+		t.Skip("execution tracing is already enabled")
```
Comment (ajgajg1134) [inline contrib/database/sql/exec_trace_test.go:30]: hmm what could cause this test to be skipped here? Is this just a guard against other tests possibly leaving runtime tracing on?

File: contrib/database/sql/exec_trace_test.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	}
+
+	// In-memory server & client which discards everything, to avoid
+	// slowness from unnecessary network I/O
+	s, c := httpmem.ServerAndClient(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
+	defer s.Close()
+	tracer.Start(tracer.WithHTTPClient(c), tracer.WithLogStartup(false))
```
[the remark this answers is not in the document]
Comment (ajgajg1134) [reply contrib/database/sql/exec_trace_test.go:37]: oh jk I see you called tracer.Stop() later... is it important for that call to happen there? If not I think we should prefer the `defer` to make it obvious to any readers that the tracer can't get leaked here

**Review by ajgajg1134 — CHANGES_REQUESTED** [review]
Overall looks good to me, it definitely adds a _bit_ of complication to the database/sql contrib code, but unfortunately I can't see a better way to deal with the driver.ErrSkip business that causes this contrib to be so complicated.

File: contrib/database/sql/exec_trace_test.go (hunk 1 above)
Comment (nsrip-dd) [reply contrib/database/sql/exec_trace_test.go:36]: This is a guard against somebody running the tests with tracing already enabled, like `go tool test -trace=trace.out`.

File: contrib/database/sql/exec_trace_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nsrip-dd) [reply contrib/database/sql/exec_trace_test.go:43]: Now that I think about it more, yeah, I think all I want to know is that APM tracing was enabled while execution tracing was also enabled. So deferring the tracer.Stop should be fine 👍

File: contrib/database/sql/tx.go [hunk 3]
```diff
@@ -20,18 +23,47 @@ type tracedTx struct {
...
+//
+// This is intended for cases where a span would normally be created after an
+// operation, where the operation may have been skipped and a span would be
+// noisy. Execution trace tasks must cover the actual duration of the operation
+// and can't be altered after the fact.
+func startTraceTask(ctx context.Context, name string) (context.Context, func()) {
+	if !trace.IsEnabled() {
+		return ctx, noopTaskEnd
+	}
+	// TODO: we could consider skipping the name. Normally we would try to
```
[the remark this answers is not in the document]
Comment (nsrip-dd) [reply contrib/database/sql/tx.go]: Yeah, as I added more testing I realized I was relying on the names. The stuff we're working on doesn't directly rely on them, but now that the test does I think we should keep them as-is. I'll remove this comment.

File: internal/trace_context.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+type executionTracedKey struct{}
+
+// WithExecutionTraced marks ctx as being associated with an execution trace
+// task. It is assumed that ctx already contains a trace task. The caller is
+// responsible for ending the task.
+//
+// This is intended for a specific case where the database/sql contrib package
+// only creates spans *after* an operation, in case the operation was
+// unvailable, and thus execution trace tasks tied to the span only capture the
```
[the remark this answers is not in the document]
Comment (nsrip-dd) [reply internal/trace_context.go]: Fixed, thanks!

File: contrib/database/sql/exec_trace_test.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023 Datadog, Inc.
+
+// TODO: skip if Go >= 1.21, until gotraceui handles those format changes?
```
[the remark this answers is not in the document]
Comment (nsrip-dd) [reply contrib/database/sql/exec_trace_test.go]: To be safe, I'll skip the tests for Go 1.21 and later and make a note to check back once gotraceui supports it. We'll still have coverage for older Go versions due to our support policy, so this stuff should still be covered in our CI.

**Review by nsrip-dd — COMMENTED** [review]
Thanks for the review Andrew!

**Review by ajgajg1134 — APPROVED** [review]
Looks good to me!

**Pull request merged**